helix-view/src/handlers/lsp.rs

```diff
@@ -1 +1,241 @@
+                                    }
+                                })?;
+                            }
+                        }
+                    }
+                }
+            }
+
+            return Ok(());
+        }
+
+        if let Some(ref changes) = workspace_edit.changes {
+            log::debug!("workspace changes: {:?}", changes);
+            for (i, (uri, text_edits)) in changes.iter().enumerate() {
+                let text_edits = text_edits.to_vec();
+                apply_edits(self, uri, None, text_edits).map_err(|kind| ApplyEditError {
+                    kind,
+                    failed_change_idx: i,
+                })?;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn apply_document_resource_op(&mut self, op: &lsp::ResourceOp) -> std::io::Result<()> {
+        use lsp::ResourceOp;
+        use std::fs;
+        match op {
+            ResourceOp::Create(op) => {
+                let path = op.uri.to_file_path().unwrap();
+                let ignore_if_exists = op.options.as_ref().map_or(false, |options| {
+                    !options.overwrite.unwrap_or(false) && options.ignore_if_exists.unwrap_or(false)
+                });
+                if !ignore_if_exists || !path.exists() {
+                    // Create directory if it does not exist
+                    if let Some(dir) = path.parent() {
+                        if !dir.is_dir() {
+                            fs::create_dir_all(dir)?;
+                        }
+                    }
+
+                    fs::write(&path, [])?;
+                    self.language_servers.file_event_handler.file_changed(path);
+                }
+            }
+            ResourceOp::Delete(op) => {
+                let path = op.uri.to_file_path().unwrap();
+                if path.is_dir() {
+                    let recursive = op
+                        .options
+                        .as_ref()
+                        .and_then(|options| options.recursive)
+                        .unwrap_or(false);
+
+                    if recursive {
+                        fs::remove_dir_all(&path)?
+                    } else {
+                        fs::remove_dir(&path)?
+                    }
+                    self.language_servers.file_event_handler.file_changed(path);
+                } else if path.is_file() {
+                    fs::remove_file(&path)?;
+                }
+            }
+            ResourceOp::Rename(op) => {
+                let from = op.old_uri.to_file_path().unwrap();
+                let to = op.new_uri.to_file_path().unwrap();
+                let ignore_if_exists = op.options.as_ref().map_or(false, |options| {
+                    !options.overwrite.unwrap_or(false) && options.ignore_if_exists.unwrap_or(false)
+                });
+                if !ignore_if_exists || !to.exists() {
+                    self.move_path(&from, &to)?;
+                }
+            }
+        }
+        Ok(())
+    }
+}
```

Comment:
this was moved here from helix_term and responsible for most of the diff. I needed to call this from editor.rs. I think this never belonged in helix-term to begin with.